### PR TITLE
Enable wandb logging for Ray Tune HPO runs

### PR DESCRIPTION
**Pull request by krfricke**
Currently wandb logging does not work with Ray Tune, as each process tries to call `wandb.log` without initializing wandb first.

This PR fixes this by wrapping the objective in Ray Tune's `wandb_mixin`, which makes sure each trial initializes wandb and logs to it.

### CODE REVIEW

**Comment by julien-c** [comment]
making sure that @borisdayma sees this PR.

**Comment by borisdayma** [comment]
Thanks, I've not played too much with ray tune yet but there seems to be 2 ways to integrate through Ray Tune libraries as per [the docs](https://docs.wandb.com/library/integrations/ray-tune).

However ideally, the setup, logging, etc would be handled directly by `Trainer` existing functions for clarity and concision (and also to support all existing loggers). Handling multiple logging runs should be done within `hyperparameter_search` if possible.

Could the setup methods be wrapped in a new function and called during the search, in order to avoid duplicating the same logic.
For wandb, forcing a new run just requires `wandb.init(reinit=True)` so it works both in notebooks and scripts.

Note: Use this argument **only** while using `hyperparameter_search` as users can currently call manually `wandb.init` before (for example when using pytorch-lightning, sweeps, or keras + huggingface), making the call within the `Trainer` a "noop" (because it does not have `reinit=True`).

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -898,7 +900,18 @@ def _objective(trial):
                 from ray.tune import CLIReporter
 
                 kwargs["progress_reporter"] = CLIReporter(metric_columns=["objective"])
-            analysis = tune.run(_objective, config=self.hp_space(None), num_samples=n_trials, **kwargs)
+            config = self.hp_space(None)
+
+            # Enable per-trial wandb logging
+            if is_wandb_available() and os.getenv("WANDB_DISABLED", "false") != "true":
```
Comment (sgugger) [inline src/transformers/trainer.py:906]: Same comment as on #6747. Should it be part of the same helper function?

**Comment by krfricke** [comment]
Thanks for your comments.

@borisdayma, simply moving the logger setup to `train()` would do the trick in any case, as it is called from the hyperparameter search methods. This should also work for Optuna, not only for Ray Tune.

I created a PR for that here: #6850. Is this what you meant?

**Comment by krfricke** [comment]
Closed in favor of #6850.

**Pull request closed without merging**